kv/kvserver: TestStoreMetrics failed

[(kv/kvserver).TestStoreMetrics failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2658618&tab=buildLog) on [master@dc75a420832fadce803482b0bcbe4bec5a3623a6](https://github.com/cockroachdb/cockroach/commits/dc75a420832fadce803482b0bcbe4bec5a3623a6):

```
=== RUN   TestStoreMetrics
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreMetrics834411500
    test_log_scope.go:74: use -show-logs to present logs inline
    client_metrics_test.go:289: livebytes for store cluster_id:56793f21-7c7e-47b7-8443-f604d48ddd2c node_id:3 store_id:3: gauge 20 != computed 41
    client_metrics_test.go:289: keybytes for store cluster_id:56793f21-7c7e-47b7-8443-f604d48ddd2c node_id:3 store_id:3: gauge 14 != computed 29
    client_metrics_test.go:289: valbytes for store cluster_id:56793f21-7c7e-47b7-8443-f604d48ddd2c node_id:3 store_id:3: gauge 6 != computed 12
    client_metrics_test.go:289: livecount for store cluster_id:56793f21-7c7e-47b7-8443-f604d48ddd2c node_id:3 store_id:3: gauge 1 != computed 2
    client_metrics_test.go:289: keycount for store cluster_id:56793f21-7c7e-47b7-8443-f604d48ddd2c node_id:3 store_id:3: gauge 1 != computed 2
    client_metrics_test.go:289: valcount for store cluster_id:56793f21-7c7e-47b7-8443-f604d48ddd2c node_id:3 store_id:3: gauge 1 != computed 2
    client_metrics_test.go:289: verifyStats failed, aborting test.
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreMetrics834411500
--- FAIL: TestStoreMetrics (1.16s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreMetrics PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreMetrics.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreMetrics failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2704365&tab=buildLog) on [master@70740041e8aa6eaa10ab527dc52a17ea362a6018](https://github.com/cockroachdb/cockroach/commits/70740041e8aa6eaa10ab527dc52a17ea362a6018):

```
=== RUN   TestStoreMetrics
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreMetrics386654614
    test_log_scope.go:74: use -show-logs to present logs inline
    client_metrics_test.go:324: sysbytes for store cluster_id:30386828-08a8-4582-8032-958ffaa1a892 node_id:3 store_id:3: gauge 1364 != computed 2439
    client_metrics_test.go:324: syscount for store cluster_id:30386828-08a8-4582-8032-958ffaa1a892 node_id:3 store_id:3: gauge 7 != computed 11
    client_metrics_test.go:324: verifyStats failed, aborting test.
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreMetrics386654614
--- FAIL: TestStoreMetrics (4.73s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreMetrics PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreMetrics.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreMetrics failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2779691&tab=buildLog) on [master@77a21a4c711c41a25a5717940e82c06488d5599e](https://github.com/cockroachdb/cockroach/commits/77a21a4c711c41a25a5717940e82c06488d5599e):

```
=== RUN   TestStoreMetrics
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreMetrics687664764
    test_log_scope.go:74: use -show-logs to present logs inline
    client_metrics_test.go:324: sysbytes for store cluster_id:8e1e9ea9-c158-438f-a0a4-83f3bd7e2c25 node_id:3 store_id:3: gauge 2495 != computed 2310
    client_metrics_test.go:324: syscount for store cluster_id:8e1e9ea9-c158-438f-a0a4-83f3bd7e2c25 node_id:3 store_id:3: gauge 12 != computed 11
    client_metrics_test.go:324: verifyStats failed, aborting test.
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreMetrics687664764
--- FAIL: TestStoreMetrics (1.25s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreMetrics PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreMetrics.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/client_metrics_test.go
+++ b/pkg/kv/kvserver/client_metrics_test.go
@@ -13,6 +13,8 @@ package kvserver_test
 import (
 	"context"
 	"fmt"
+	"strconv"
+	"sync"
 	"testing"
 	"time"
 
@@ -54,6 +56,8 @@ func checkGauge(t *testing.T, id string, g gaugeValuer, e int64) {
 func verifyStats(t *testing.T, tc *testcluster.TestCluster, storeIdxSlice ...int) {
 	t.Helper()
 	var stores []*kvserver.Store
+	var wg sync.WaitGroup
+
 	for _, storeIdx := range storeIdxSlice {
 		stores = append(stores, tc.GetFirstStoreFromServer(t, storeIdx))
 	}
@@ -70,6 +74,20 @@ func verifyStats(t *testing.T, tc *testcluster.TestCluster, storeIdxSlice ...int
 			return nil
 		})
 	}
+
+	wg.Add(len(storeIdxSlice))
+	// We actually stop *all* of the Servers. Stopping only a few is riddled
+	// with deadlocks since operations can span nodes, but stoppers don't
+	// know about this - taking all of them down at the same time is the
+	// only sane way of guaranteeing that nothing interesting happens, at
+	// least when bringing down the nodes jeopardizes majorities.
+	for _, storeIdx := range storeIdxSlice {
+		go func(i int) {
+			defer wg.Done()
+			tc.StopServer(i)
+		}(storeIdx)
+	}
+	wg.Wait()
 
 	for _, s := range stores {
 		idString := s.Ident.String()
@@ -106,6 +124,11 @@ func verifyStats(t *testing.T, tc *testcluster.TestCluster, storeIdxSlice ...int
 
 	if t.Failed() {
 		t.Fatalf("verifyStats failed, aborting test.")
+	}
+
+	// Restart all Stores.
+	for _, storeIdx := range storeIdxSlice {
+		require.NoError(t, tc.RestartServer(storeIdx))
 	}
 }
 
@@ -228,25 +251,36 @@ func TestStoreMetrics(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
-	tc := testcluster.StartTestCluster(t, 3,
-		base.TestClusterArgs{
-			ReplicationMode: base.ReplicationManual,
-			ServerArgs: base.TestServerArgs{
-				StoreSpecs: []base.StoreSpec{
-					{
-						InMemory: true,
-						// Specify a size to trigger the BlockCache in Pebble.
-						Size: base.SizeSpec{
-							InBytes: 1 << 20,
-						},
-					},
-				},
-				Knobs: base.TestingKnobs{
-					Store: &kvserver.StoreTestingKnobs{
-						DisableRaftLogQueue: true,
+	stickyEngineRegistry := server.NewStickyInMemEnginesRegistry()
+	defer stickyEngineRegistry.CloseAllStickyInMemEngines()
+	const numServers int = 3
+	stickyServerArgs := make(map[int]base.TestServerArgs)
+	for i := 0; i < numServers; i++ {
+		stickyServerArgs[i] = base.TestServerArgs{
+			StoreSpecs: []base.StoreSpec{
+				{
+					InMemory:               true,
+					StickyInMemoryEngineID: strconv.FormatInt(int64(i), 10),
+					// Specify a size to trigger the BlockCache in Pebble.
+					Size: base.SizeSpec{
+						InBytes: 1 << 20,
 					},
 				},
 			},
+			Knobs: base.TestingKnobs{
+				Server: &server.TestingKnobs{
+					StickyEngineRegistry: stickyEngineRegistry,
+				},
+				Store: &kvserver.StoreTestingKnobs{
+					DisableRaftLogQueue: true,
+				},
+			},
+		}
+	}
+	tc := testcluster.StartTestCluster(t, numServers,
+		base.TestClusterArgs{
+			ReplicationMode:   base.ReplicationManual,
+			ServerArgsPerNode: stickyServerArgs,
 		})
 	defer tc.Stopper().Stop(ctx)
 
@@ -302,9 +336,7 @@ func TestStoreMetrics(t *testing.T) {
 	// Verify stats after addition.
 	verifyStats(t, tc, 1, 2)
 	checkGauge(t, "store 0", tc.GetFirstStoreFromServer(t, 0).Metrics().ReplicaCount, initialCount+1)
-
-	tc.TransferRangeLeaseOrFatal(t, desc, tc.Target(1))
-	tc.RemoveVotersOrFatal(t, key, tc.Target(0))
+	tc.RemoveLeaseHolderOrFatal(t, desc, tc.Target(0), tc.Target(1))
 	testutils.SucceedsSoon(t, func() error {
 		_, err := tc.GetFirstStoreFromServer(t, 0).GetReplica(desc.RangeID)
 		if err == nil {

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -847,6 +847,27 @@ func (tc *TestCluster) TransferRangeLeaseOrFatal(
 	}
 }
 
+// RemoveLeaseHolderOrFatal is a convenience version of TransferRangeLease and RemoveVoter
+func (tc *TestCluster) RemoveLeaseHolderOrFatal(
+	t testing.TB,
+	rangeDesc roachpb.RangeDescriptor,
+	src roachpb.ReplicationTarget,
+	dest roachpb.ReplicationTarget,
+) {
+	testutils.SucceedsSoon(t, func() error {
+		if err := tc.TransferRangeLease(rangeDesc, dest); err != nil {
+			return err
+		}
+		if _, err := tc.RemoveVoters(rangeDesc.StartKey.AsRawKey(), src); err != nil {
+			if strings.Contains(err.Error(), "to remove self (leaseholder)") {
+				return err
+			}
+			t.Fatal(err)
+		}
+		return nil
+	})
+}
+
 // MoveRangeLeaseNonCooperatively is part of the TestClusterInterface.
 func (tc *TestCluster) MoveRangeLeaseNonCooperatively(
 	rangeDesc roachpb.RangeDescriptor, dest roachpb.ReplicationTarget, manual *hlc.HybridManualClock,

